design/one-pager-composition-revisions.md

```diff
@@ -0,0 +1,300 @@
+
+```go
+// A CompositionFetcher fetches an appropriate Composition for the supplied
+// composite resource.
+type CompositionFetcher interface {
+  Fetch(ctx context.Context, cr resource.Composite) (*v1.Composition, error)
+}
+
+// r.composition is a CompositionFetcher.
+comp, err := r.composition.Fetch(ctx, cr)
+if err != nil {
+  return reconcile.Result{RequeueAfter: shortWait}, nil
+}
+```
+
+Each XR reconciler will default to using a `CompositionFetcher` that defaults to
+the same logic we use today; i.e. a `client.Client` `Get`. When Crossplane is
+started with a feature flag such as `--enable-alpha-composition-revisions` this
+default `CompositionFetcher` implementation will be replaced with one that
+instead:
+
+1. Fetches the `Composition` from the API server.
+2. Selects the appropriate `CompositionRevision`.
+3. Returns the selected `CompositionRevision` converted to a `Composition`.
+
+Converting the `CompositionRevision` to a `Composition` allows us to introduce
+support for revisions with minimal changes to the XR reconcile logic. This will
+allow us to become comfortable with a `v1alpha1` iteration of the new API type
+before we commit to it. If and when we are comfortable with the `v1alpha1`
+`CompositionRevision` API we can promote it to `v1beta1` and have it on by
+default. At this point the XR reconcile logic would be updated to deal directly
+with the `CompositionRevision` API type, without any conversion back to the
+`Composition` type. Notably this would involve migrating the patch logic that is
+currently defined as methods on the `Composition` type to being methods on the
+`CompositionRevision` type.
+
+## Future Considerations
+
+Two key future considerations are garbage collection of revisions, and allowing
+platform engineers to enforce a `compositionUpdatePolicy`.
+
+### Garbage Collection
+
+'Revisions' are an established pattern in the Kubernetes ecosystem per the
+"Prior Art" section below. It's common for controllers that create revisions to
+garbage collect older revisions in order to avoid bloating the API server with
+unused objects. Typically the latest N revisions of a type are kept, and
+anything older is automatically garbage collected. Most revision implementations
+are designed such that only one revision will be 'active' at any one time. For
+example there is only ever one active `ProviderRevision` for each Crossplane
+`Provider` resource. This means it's generally fairly safe and easy to garbage
+collect old revisions; you need only avoid garbage collecting the active one.
+
+The design proposed by this document requires a more nuanced approach to
+garbage collecting old revisions, because many revisions may be in use at one
+time. The policy would likely need to be "garbage collect revisions older than N
+if they're not in use", which would require the garbage collector to keep track
+of which revisions are in use across an arbitrary set of XR types. This document
```

Comment:
I wonder if this reference counting could be handled using built-in functionality (namely, `finalizers`). It would require adding a finalizer for each XR that utilized a `CompositionRevision`, but it would make the deletion logic simpler as we would simply always try to delete a `CompositionRevision` that was outside of the GC window, and k8s would be responsible for making sure it stuck around. Also has the nice side-effect of exposing what XRs are blocking deletion of a given `CompositionRevision`.

Reply:
Is this pattern of using many finalizers common? It always seems like a bit of a misuse of finalizers to me when I hear it proposed - probably because I think of finalizers as being one-per-reconciler rather than one-per-related-object.

Reply:
@negz I'm not sure, it does feel somewhat unconventional, but based on the [description of the purpose of a finalizer](https://kubernetes.io/blog/2021/05/14/using-finalizers-to-control-deletion/#understanding-finalizers), it doesn't necessarily feel like a misuse:
> Finalizers are keys on resources that signal pre-delete operations. They control the garbage collection on resources, and are designed to alert controllers what cleanup operations to perform prior to removing a resource. However, they don’t necessarily name code that should be executed; finalizers on resources are basically just lists of keys much like annotations. Like annotations, they can be manipulated.